lib/vector-core/src/event/metric/data.rs

```diff
@@ -83,6 +90,28 @@ impl MetricData {
                 (Some(t), None) | (None, Some(t)) => Some(t),
                 (Some(t1), Some(t2)) => Some(t1.max(t2)),
             };
+
+            let delta_t = self
+                .timestamp
+                .and_then(|ts| {
+                    other
+                        .timestamp
+                        .map(|other_ts| (ts.timestamp_millis() - other_ts.timestamp_millis().abs()))
+                })
+                .unwrap_or(0) as u64;
+
+            self.interval_ms = match (self.interval_ms, other.interval_ms) {
+                // If either interval is None discard the other
+                (_, None) | (None, _) => None,
+                // If metrics timestamps are within their interval range (should be the usual case) we use the longest interval
+                (Some(i1), Some(i2)) => {
+                    if delta_t < i1.max(i2) {
+                        Some(i1.max(i2))
+                    } else {
+                        Some(i1 + i2)
+                    }
+                }
```

Comment:
I'm not sure this logic covers all the cases. Consider if `self` has timestamp 100 and interval 1, and `other` has timestamp 110 and interval 20. `delta_t` will be 10, and the max interval is 20, but the two intervals are disjoint.